volatility3/framework/plugins/windows/svcscan.py

```diff
@@ -232,28 +232,40 @@ def _get_service_binary_map(
             for service_key in services
         }
 
-    def _generator(self):
-        kernel = self.context.modules[self.config["kernel"]]
+    def enumerate_vista_or_later_header(
+        self, service_table_name, service_binary_dll_map, proc_layer_name, offset
+    ):
+        if offset % 8:
+            return
 
-        service_table_name = self.create_service_table(
-            self.context, kernel.symbol_table_name, self.config_path
+        service_header = self.context.object(
+            service_table_name + constants.BANG + "_SERVICE_HEADER",
+            offset=offset,
+            layer_name=proc_layer_name,
         )
 
-        # Building the dictionary ahead of time is much better for performance
-        # vs looking up each service's DLL individually.
-        services_key = self._get_service_key(kernel)
-        service_binary_dll_map = (
-            self._get_service_binary_map(services_key)
-            if services_key is not None
-            else {}
-        )
+        if not service_header.is_valid():
+            return
+
+        # since we walk the s-list backwards, if we've seen
+        # an object, then we've also seen all objects that
+        # exist before it, thus we can break at that time.
+        for service_record in service_header.ServiceRecord.traverse():
+            service_info = service_binary_dll_map.get(
+                service_record.get_name(),
+                ServiceBinaryInfo(
+                    renderers.UnreadableValue(), renderers.UnreadableValue()
+                ),
+            )
+            yield self.get_record_tuple(service_record, service_info)
+
+    def service_scan(self, service_table_name, service_binary_dll_map, filter_func):
```

Comment:
Same again, unless it's `run` or `generator` or something that's private/will only ever be used by that specific plugin, it should be parameterized with at least `context` and exposed as a `classmethod` (to make sure self doesn't work accidentally).